### ISSUE TITLE
vite-plugin-dynamic-import works in dev but not in build/production

### ISSUE THREAD

**Issue by stereokai** [caoxiemeihao/vite-plugins#29]
Using [vite-plugin-dynamic-import](https://github.com/caoxiemeihao/vite-plugins/blob/main/packages/dynamic-import), I have the following setup, which works in `vite dev` environment:

Main module → static import module A → dynamic import module B → static import module C

Initially, when running `vite build` I received this error:
```
[vite:esbuild-transpile] Transform failed with 1 error:
assets/index.80ab11b6.js:328:14: ERROR: Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2019", "firefox78", "safari13.1")

Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2019", "firefox78", "safari13.1")
326|  };
327|
328|  const chart = await __variableDynamicImportRuntime0__(`@/renderers/${renderer}/${renderer}.js`);
```

But I was able to fix it with the following in vite config:
```js
build: {
    target: "esnext",
}
```

At this point, the app builds correctly and loads using `vite preview`. I can see the dynamic imports appearing in the network tab. 

The problem is:
1) Dynamic imports are loaded but not executed (code never runs)
2) Code execution stops after the above line 328, because the expression ``await __variableDynamicImportRuntime0__(`@/renderers/${renderer}/${renderer}.js`)`` is never resolved
3) It works if I switch the `await` expression to a promise ``import(`@/renderers/${renderer}/${renderer}.js`).then(/* ... */)``

What am I missing here? Thanks!

**Comment by caoxiemeihao** [comment]
Can you provide the complete code of **import(`@/renderers/${renderer}/${renderer}.js`)**?

It is best to provide a minimum runnable demo 😄

**Comment by stereokai** [comment]
Hi, thanks for the quick reply! Do you mean, the code of what's inside `@/renderers/${renderer}/${renderer}.js`?

**Comment by caoxiemeihao** [comment]
The file where this code **import(`@/renderers/${renderer}/${renderer}.js`)** is located

**Comment by stereokai** [comment]
@caoxiemeihao You can see it here: https://github.com/stereokai/multi-charts-comparison/blob/main/renderers/Echarts/Echarts.js

The import path to that file is:
app.js → graphs.js → Echarts.js

If this doesn't help, I can send you a minimum reproducible demo in about 1.5-2 hours

**Comment by stereokai** [comment]
@caoxiemeihao https://github.com/stereokai/multi-charts-comparison/tree/example_for_caoxiemeihao
I created a minimum reproducible demo, but confusingly, in this demo, everything works as expected... Something about the code structure in the full project is causing dynamic imports to not execute the async loaded modules in production and not resolve the `await`ed promise... But weirdly with `import().then()` it doesn't happen and everything resolves and executes as expected.

 I don't want to overload you with information, but maybe it'll help shed a light on what I'm (obviously) doing wrong – even weirder is that the rollup configuration solves the original problem and makes the production version work just like the dev version:

```js
      output: {
        manualChunks(id) {
          id = id.replace(__dirname, "").toLowerCase();

          if (id.includes("workerpool")) {
            return "workerpool";
          }
          if (id.includes("echarts@") || id.includes("zrender@")) {
            return "echarts";
          }
          if (id.includes("echarts")) {
            return "echarts.renderer";
          }
          if (id.includes("node_modules")) {
            return "vendor";
          }
        },
```

**Comment by caoxiemeihao** [comment]
We have jet lag. I just came back from dinner;  
Now I'm going to look at your problem 😄

**Comment by caoxiemeihao** [comment]
You can see the js request in Network

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/26263658/167405182-a21f97d6-cecc-497d-9a19-4bb74e750fd6.png">

See the complete code in Response

<img width="1051" alt="image" src="https://user-images.githubusercontent.com/26263658/167405299-2e2f93ee-1bed-4a4d-a7df-19dcfc797fce.png">

This code should be in line with expectations

```js
import {
  addChannels,
  channels,
  regenerateAllChannels,
  removeChannels,
} from "/models/state.js";
import { renderer } from "/router.js";
export { default as EVENTS } from "/Graphs/graphEvents.js";
export { renderer };

const chart = await __variableDynamicImportRuntime0__(`@/renderers/${renderer}/${renderer}.js`);

let prevSamplesPerChannel = 0;
let prevSamplesPerSecond = 0;
let timeSeries;

export function on(...args) {
  chart.on(...args);
}

export function initGraph(samplesPerChannel, samplesPerSecond) {
  onTotalSamplesChange(samplesPerChannel, samplesPerSecond);
  chart.init();
  setChartData(() => regenerateAllChannels(samplesPerChannel));
}

function updateChart(dataset) {
  chart.update(dataset, timeSeries);
}

function setChartData(dataOperation) {
  chart.showLoading();
  dataOperation()
    .then(updateChart)
    .catch((err) => {
      console.log("Operation error", err);
    })
    .finally(() => {
      chart.hideLoading();
    });
}

function onTotalSamplesChange(samplesPerChannel, samplesPerSecond) {
  prevSamplesPerChannel = samplesPerChannel;
  prevSamplesPerSecond = samplesPerSecond;

  timeSeries = generateTimeseries(samplesPerChannel, samplesPerSecond);
}

function generateTimeseries(totalSamples, samplesPerSecond) {
  const baseDate = new Date();
  baseDate.setDate(baseDate.getDate() - 1);
  const timestamp = baseDate.setHours(22);

  const samplesPerTimestampInSeconds = 1 / samplesPerSecond;
  const samplesPerTimestampInMilliseconds = samplesPerTimestampInSeconds * 1000;

  const timestamps = new Array(totalSamples);
  for (let i = 0; i < totalSamples; i++) {
    timestamps[i] = timestamp + i * samplesPerTimestampInMilliseconds;
  }
  return timestamps;
}

export function onSettingChange(
  numberOfChannels,
  samplesPerSecond,
  totalSamples
) {
  const samplesPerChannel = Math.floor(totalSamples / numberOfChannels);
  const channelsChange = numberOfChannels - channels.length;
  let operation;

  // Channels removed
  if (channelsChange < 0) {
    removeChannels(channelsChange);
    updateChart();
    return;
  }

  // No. of samples per channel changed
  if (
    samplesPerChannel !== prevSamplesPerChannel ||
    samplesPerSecond !== prevSamplesPerSecond
  ) {
    onTotalSamplesChange(samplesPerChannel, samplesPerSecond);
    operation = () => regenerateAllChannels(samplesPerChannel);
  }

  // Channels added
  if (channelsChange > 0) {
    operation = () => addChannels(channelsChange, samplesPerChannel);
  }

  setChartData(operation);
}

// --------- vite-plugin-dynamic-import ---------

function __variableDynamicImportRuntime0__(path) {
  switch (path) {
    case '@/renderers/Echarts/Echarts.js':
      return import('/renderers/Echarts/Echarts.js');

    case '@/renderers/Echarts/echartsDefaults.js':
      return import('/renderers/Echarts/echartsDefaults.js');

    case '@/renderers/Echarts/echartsEvents.js':
      return import('/renderers/Echarts/echartsEvents.js');

    case '@/renderers/Echarts/echartsOptionsBuilder.js':
      return import('/renderers/Echarts/echartsOptionsBuilder.js');

    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
  }
}
```

**Comment by caoxiemeihao** [comment]
When set `build.minify = false`, will generated the blow code.

```js
// ...omit
const chart = await __variableDynamicImportRuntime0__(`@/renderers/${renderer}/${renderer}.js`);
// ...omit

// --------- vite-plugin-dynamic-import ---------

function __variableDynamicImportRuntime0__(path) {
  switch (path) {
    case '@/renderers/Echarts/Echarts.js':
      return __vitePreload(() => import('./echarts.renderer.b5602ebb.js').then(function (n) { return n.E; }),true?["assets/echarts.renderer.b5602ebb.js","assets/workerpool.f49a4ca2.js","assets/echarts.b7a16009.js","assets/vendor.efe0d25d.js","assets/vendor.3d0ef3f1.css"]:void 0);

    case '@/renderers/Echarts/echartsDefaults.js':
      return __vitePreload(() => import('./echarts.renderer.b5602ebb.js').then(function (n) { return n.e; }),true?["assets/echarts.renderer.b5602ebb.js","assets/workerpool.f49a4ca2.js","assets/echarts.b7a16009.js","assets/vendor.efe0d25d.js","assets/vendor.3d0ef3f1.css"]:void 0);

    case '@/renderers/Echarts/echartsEvents.js':
      return __vitePreload(() => import('./echarts.renderer.b5602ebb.js').then(function (n) { return n.f; }),true?["assets/echarts.renderer.b5602ebb.js","assets/workerpool.f49a4ca2.js","assets/echarts.b7a16009.js","assets/vendor.efe0d25d.js","assets/vendor.3d0ef3f1.css"]:void 0);

    case '@/renderers/Echarts/echartsOptionsBuilder.js':
      return __vitePreload(() => import('./echarts.renderer.b5602ebb.js').then(function (n) { return n.g; }),true?["assets/echarts.renderer.b5602ebb.js","assets/workerpool.f49a4ca2.js","assets/echarts.b7a16009.js","assets/vendor.efe0d25d.js","assets/vendor.3d0ef3f1.css"]:void 0);

    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
  }
}

```

**Comment by caoxiemeihao** [comment]
@stereokai Hey! 👋

I created an PR in you project for fix the problem 👉 https://github.com/stereokai/multi-charts-comparison/pull/1/files

**Comment by stereokai** [comment]
@caoxiemeihao That's incredible! Thank you so much! So kind of you! Where did you get the idea from? And how can I return the favor?

**Comment by caoxiemeihao** [comment]
@stereokai You're welcome. You can treat me as a good friend!

**Comment by stereokai** [comment]
Cool, let's do that :)

Do you have any source of information that connects module loading/execution/async resolution with document state `"interactive"`?

**Comment by caoxiemeihao** [comment]
> My English is not good. I hope you can understand what I mean 😅

---

In `vite serve` phase, we have below import-chain

> index.html > /main.js > app.js > Graphs.js

The `Graphs.js` contains `import()`, time line after three times the import-HTPP requests and this is the Promise.  
When the whole code of board, `"document.readyState"` has a value of `"complete"`

After `vite build` execution, the import-chain will become like this

> index.html > /assets/index.hash256.js

`Graphs.js` of the code will be bound to the `index.hash256.js`, time online only once the import-HTTP requests.  
When the whole code of board, `"document.readyState"` has a value of `"interactive"`

---

**To sum up, the results of my local test 😄**

Ref: https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState#value

**Comment by stereokai** [comment]
Yes!!! Fantastic! Thanks you very much! 😄
https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState#readystatechange_as_an_alternative_to_domcontentloaded_event
